redskyctl/internal/commander/reader.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2020 GramLabs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package commander
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	redskyv1alpha1 "github.com/redskyops/redskyops-controller/api/v1alpha1"
+	redskyv1beta1 "github.com/redskyops/redskyops-controller/api/v1beta1"
+	"github.com/redskyops/redskyops-controller/internal/controller"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// ResourceReader helps properly decode Kubernetes resources on the CLI. It is meant to be a
+// lighter weight alternative to the cli-runtime resource.Builder.
+type ResourceReader struct {
+	Scheme           *runtime.Scheme
+	PreferredVersion runtime.GroupVersioner
+}
+
+// NewResourceReader returns a new resource reader for the supplied byte stream.
+func NewResourceReader() *ResourceReader {
+	rr := &ResourceReader{
+		Scheme:           runtime.NewScheme(),
+		PreferredVersion: OnlyVersion,
+	}
+
+	// Always add our types
+	_ = redskyv1beta1.AddToScheme(rr.Scheme)
+	_ = redskyv1alpha1.AddToScheme(rr.Scheme)
+
+	// Allow single experiments to target an experiment list
+	_ = addExperimentListConversions(rr.Scheme)
+
+	return rr
+}
+
+// ReadInto decodes the supplied byte stream into the target runtime object. The default values
+// and type information of the target object will be populated.
+func (r *ResourceReader) ReadInto(reader io.ReadCloser, target runtime.Object) error {
```

Comment:
I think only issue I have is `ReadInto(runtime.Object)` seems too generic when we only support our objects. 
Should we extend the reader to support all kube objects or rename `ReadInto` to something more specific ( `ReadRedSkyObject` or something like that ).

Reply:
We only support our objects by default to cut down on boilerplate, nothing stops you from doing a `_ = clientgoscheme.AddToScheme(rr.Scheme)` and then using that `ResourceReader` to read a core Kube object.